lib/Solver/CachingSolver.cpp

```diff
@@ -34,12 +34,17 @@ class CachingSolver : public SolverImpl {
   
   struct CacheEntry {
     CacheEntry(const ConstraintSet &c, ref<Expr> q)
-      : constraints(c), query(q) {}
+        : constraints(c.begin(), c.end()), query(q) {
+      // Order constraints
+      std::sort(
+          constraints.begin(), constraints.end(),
+          [&](const ref<Expr> &a, const ref<Expr> &b) { return *a < *b; });
+    }
 
     CacheEntry(const CacheEntry &ce)
       : constraints(ce.constraints), query(ce.query) {}
-    
-    ConstraintSet constraints;
+
+    std::vector<ref<Expr>> constraints;
```

Comment:
This looks very odd too. I thought you introduced `ConstraintSet` precisely to avoid this issue. Maybe a solution would be just to `typedef ConstraintSet std::vector<ref<Expr>>` on the other hand if you want to limit control to ConstraintSet for some reason, maybe it would make sense to just have an invariant where ConstraintSets are always ordered?

I any case I find it quite strange that the constraints get ordered here in the CachingSolver.

Reply:
Well, the problem is how to order?
For the path constraints, you might want to keep the insert order of the constraints. But in the case of the caches, you want force a global order of expressions to improves the hit rate and avoid false negatives. 

I first wanted to go for `ConstraintSet` as parent class and to have `StrictlyOrderedConstraintSet` and `UnorderedConstraintSet` as potential subclasses but i  thought that might be simpler in the first place.